problem4/problem4.java

```diff
@@ -0,0 +1,78 @@
+class DiningPhilosophers {
+    private final Object licenseToEat = new Object();
+    private final boolean[] isTaken = new boolean[5];
+    private final LinkedList<Integer> queue = new LinkedList<>();
+
+    public DiningPhilosophers() {
+        for (int i = 0; i < 5; i++) {
+            isTaken[i] = false;
+        }
+    }
+
+    public void wantsToEat(int philosopher,
+                           Runnable pickLeftFork,
+                           Runnable pickRightFork,
+                           Runnable eat,
+                           Runnable putLeftFork,
+                           Runnable putRightFork) throws InterruptedException {
+        int leftFork = philosopher;
+        int rightFork = (philosopher + 1) % 5;
+
+        synchronized (licenseToEat) {
+            queue.addLast(philosopher);
+
+            // Wait until it's this philosopher's turn, forks are available, and precedence over neighbors is ensured
+            while (!canPickUpForks(philosopher, leftFork, rightFork)) {
+                licenseToEat.wait();
+            }
+
+            // Pick up forks
+            isTaken[leftFork] = true;
+            isTaken[rightFork] = true;
+            queue.removeFirstOccurrence(philosopher);
+        }
+
+        // Execute the actions outside the lock to avoid blocking other threads
+        synchronized (this) {
+            pickLeftFork.run();
+            pickRightFork.run();
+            eat.run();
+            putLeftFork.run();
+            putRightFork.run();
+        }
+        
+        synchronized (licenseToEat) {
+            // Release forks
+            isTaken[leftFork] = false;
+            isTaken[rightFork] = false;
+            // Notify all waiting philosophers
+            licenseToEat.notifyAll();
```

Comment:
You could have each philosopher wait on their own object. Then, instead of notifying all here, you could notify just the philosopher that is at the top of the queue + another philosopher from the queue that doesn't conflict with the first. Not sure if this would be a better approach, but I think it decreases the number of times threads would be acquiring licenseToEat.